packages/data/src/components/use-dispatch/use-dispatch.js

```diff
@@ -0,0 +1,20 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+
+const useDispatch = ( storeName ) => {
+	const registry = useRegistry();
+	return useMemo( () => {
```

Comment:
I wonder if `useMemo` is even needed here?  Probably not?

Reply:
What's more expensive? Calling `dispatch` or the hook's overhead?

Can `registry` change without `dispatch` changing? If so, you could make line 12:

```js
const dispatch = useRegistry().dispatch;
```

And just close over that instead.

Reply:
yeah I think I'm going to drop the use of `useMemo`.  Even if `registry` changes it'll always expose `dispatch`.  So I'll probably do something like:

```js
const useDispatch = ( storeName ) => {
    const { dispatch } = useRegistry();
    return storeName === void 0 ? dispatch : dispatch( storeName );
}
```

Reply:
So calling `dispatch` on every render is fine (cheaper than the hook's overhead)?

Reply:
Essentially, `registry.dispatch` is just an object of stores mapped to action creators, so there should be little to no overhead.

Reply:
see 81ee6ee

Reply:
Can you benchmark both approaches? This will be used in so many places that slight performance differences can significantly affect the user.

Reply:
Right now the hook isn't used anywhere, so to use the performance tools we have right now I'd have to implement it in a few components that are tested with the performance tools (or there may be some other benchmarking approach I'd need to take).  If you have some ideas on how I could effectively benchmark both approaches I'm all ears.

Otherwise, I think once the hook starts getting implemented throughout the codebase we can take a look at doing some benchmarking?

Reply:
Timing `dispatch(...)` vs. `useMemo(...)` would work.